### PR TITLE
Dev pass

### PR DESCRIPTION
**Pull request by Liu-rj**
Together with #34, we get the following result on reddit.
* Config: batchsize=512, samples='10,10'

|   sec/epoch  | Sampling |  Feature Loading  |  Forward | Backward | End2End |   GPU Mem Peak (GB)   |
| ---------- | ------------- | ---- | ---- | ---- | ---- | ---- |
| Matrix | 55.782 | 0.349 | 0.595 | 2.397 | 59.126 | 2.269 |
| DGL | 41.549 | 0.343 | 0.695 | 33.794 | 76.481 | 3.634 |

### CODE REVIEW

File: src/graph.h [hunk 1]
```diff
@@ -26,14 +26,14 @@ class Graph : public torch::CustomClassHolder {
   void SetCSC(std::shared_ptr<CSC> csc);
   void SetCSR(std::shared_ptr<CSR> csr);
   void SetCOO(std::shared_ptr<COO> coo);
-  void SetData(torch::Tensor data);
+  void SetData(torch::Tensor data, std::string order = "default");
```
Comment (gpzlx1) [inline src/graph.h:29]: is direction better than order?

Comment (gpzlx1) [reply src/graph.h:29]: what's default value?

**Comment by gpzlx1** [comment]
@DanielMao1 Please profile the codes. Find out why we are slower than dgl on sampling.

File: src/graph.h (hunk 1 above)
Comment (Liu-rj) [reply src/graph.h:29]: `default` is the original data sequence without explicit edge id indexing of typical format. It is used where user just want to get the data without taking care of the data sequence order.

Comment (Liu-rj) [reply src/graph.h:29]: Same as DGL, in every matrix there is a implicit default data order with the `e_id` of corresponding format being `torch::nullopt`. For example, if the default order is `coo`, `coo_->e_ids` will be `torch::nullopt`, if we want to get the data sequence order of csc format, there will be a explicit edge id indexing.

Comment (gpzlx1) [reply src/graph.h:29]: okkk

**Comment by DanielMao1** [comment]
> @DanielMao1 Please profile the codes. Find out why we are slower than dgl on sampling.

It seems that sddmm function is slow. Maybe it should be double checked later.
https://www.notion.so/Profiling-of-pass-f2c1ef53edc24fc9a02fa5e2ff53ccac

**Comment by Liu-rj** [comment]
> > @DanielMao1 Please profile the codes. Find out why we are slower than dgl on sampling.
> 
> It seems that sddmm function is slow. Maybe it should be double checked later. https://www.notion.so/Profiling-of-pass-f2c1ef53edc24fc9a02fa5e2ff53ccac

From the observation on the profiling result, I have two further questions:
* I have noticed that DGL's kernel is `SDDMMCOO` while ours is `SDDMMCsr` right? Please check if they refers to the same operation.
* Also, I see that DGL's most costly kernel is `SpMMCsr`, then what is the equivalent operation in matrix implementation and what's its cost?

**Comment by Liu-rj** [comment]
Now that our performance loss is mainly due to sparse format selection, I think this pr can be safely merged and we can move the performance profiling to notion workspace.

**Comment by DanielMao1** [comment]
> 

The input format of SDDMM is CSC. I change it to COO and use COOSDDMM.  The end to end result shows that our sampling time is faster than dgl now.

**Comment by DanielMao1** [comment]
> > > @DanielMao1 Please profile the codes. Find out why we are slower than dgl on sampling.
> > 
> > 
> > It seems that sddmm function is slow. Maybe it should be double checked later. https://www.notion.so/Profiling-of-pass-f2c1ef53edc24fc9a02fa5e2ff53ccac
> 
> From the observation on the profiling result, I have two further questions:
> 
> * I have noticed that DGL's kernel is `SDDMMCOO` while ours is `SDDMMCsr` right? Please check if they refers to the same operation.
> * Also, I see that DGL's most costly kernel is `SpMMCsr`, then what is the equivalent operation in matrix implementation and what's its cost?

1.The input format of SDDMM is CSC. I change it to COO and use COOSDDMM. The end to end result shows that our sampling time is faster than dgl now.
2. The spmm is used in backward, which we have not implement yet.

**Review by gpzlx1 — APPROVED** [review]

**Pull request merged**